pkg/image/oci/registry_provider_test.go

```diff
@@ -0,0 +1,46 @@
+package oci
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/anchore/stereoscope/pkg/image"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_prepareReferenceOptions(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    image.RegistryOptions
+		expected []name.Option
+	}{
+		{
+			name:     "not InsecureUseHTTP",
+			input:    image.RegistryOptions{},
+			expected: nil,
+		},
+		{
+			name: "use InsecureUseHTTP",
+			input: image.RegistryOptions{
+				InsecureUseHTTP: true,
+			},
+			expected: []name.Option{name.Insecure},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			out := prepareReferenceOptions(&test.input)
+			assert.Equal(t, len(test.expected), len(out))
+			if test.expected == nil {
+				assert.Equal(t, test.expected, out)
+			} else {
```

Comment:
Question: Just double-checking my understanding of the intent here — in addition to ensuring the slice length is correct, we're adding this nil-case check for the scenario where both slices have a length of zero, but one of the slices is non-nil?

Reply:
correct, maybe I misunderstood... I have `var options []name.Option`, which is nil until it's later appended to `options = append(options, name.Insecure)` ? so this function returns `nil` or a 1-option array with the `Insecure` flag

Reply:
Yes, that's all correct. 🎉 

Maybe I could rephrase my prior comment to be more clear: We're already testing to ensure the length is the same, just a few lines up. So if we expect a zero-item slice (`nil`), and we get a zero-item slice (`nil`), that assertion will succeed.

So my question is: what is _this assertion_ checking for that hasn't already been checked?

```go
			if test.expected == nil {
				assert.Equal(t, test.expected, out)
			}
```

In a way, this is a different way of asserting something we've already asserted. But there's an exception to that statement, a particular case where that's not true. Zero-length slices can be `nil`, or they can be allocated (i.e. not `nil`) but have no array items. This assertion would alert us to a scenario where it was important to us that the function returns specifically the `nil` kind of zero-item slice, but the function instead returned an allocated zero-item slice.

I'm sort of speculating about what the intent is here, so what I'm saying is a bit contrived. In short, I think what you have is great and doesn't require a change. I'm just trying to follow along as I'm reading. 😄

Reply:
Yeah, just basically testing that we're returning `nil` as expected instead of an allocated 0-length array, also I think the `array[0]` would panic? maybe not

Reply:
Cool, thanks for the clarification.

> I think the array[0] would panic

Yep, it would, in either of the scenarios... Using a nil slice would cause a `nil pointer dereference` panic, and accessing index `0` of a non-`nil` slice with _zero items_ would cause an `index out of range` panic.